tests/unit-tests/wayland/test_lifetime_tracker.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright © 2020 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: William Wold <william.wold@canonical.com>
+ */
+
+#include "mir/wayland/wayland_base.h"
+
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+
+namespace mw = mir::wayland;
+
+using namespace testing;
+
+class MockTracker : public virtual mw::LifetimeTracker
+{
+public:
+    // Expose protected method
+    void mark_destroyed() const
+    {
+        LifetimeTracker::mark_destroyed();
+    }
+};
+
+struct MockListener
+{
+    MOCK_METHOD0(callback, void());
+};
+
+TEST(LifetimeTrackerTest, first_destroy_listener_id_is_not_zero)
+{
+    MockTracker tracker;
+    auto const id = tracker.add_destroy_listener([](){});
+    EXPECT_THAT(id.as_value(), Ne(0));
+}
+
+TEST(LifetimeTrackerTest, destroy_listener_is_called)
+{
+    MockListener listener;
+    EXPECT_CALL(listener, callback()).Times(1);
+    MockTracker tracker;
+    tracker.add_destroy_listener([&](){ listener.callback(); });
+}
+
+TEST(LifetimeTrackerTest, destroy_listener_only_called_once_when_marked_as_destroyed)
+{
+    MockListener listener;
+    EXPECT_CALL(listener, callback()).Times(1);
+    MockTracker tracker;
+    tracker.add_destroy_listener([&](){ listener.callback(); });
+    tracker.mark_destroyed();
```

Comment:
Well, I'd avoid conflating the setup and expectations:
```c++
class MockTracker ...
{
    MockListener listener;
    shared_ptr<MockTracker> tracker = make_shared<MockTracker>;

    void SetUp()
    {
        tracker->add_destroy_listener([&](){ listener.callback(); });
    }

    void TearDown()
    {
        tracker.reset();
        Mock::VerifyAndClearExpectations(&listener);
    }
...

TEST(LifetimeTrackerTest, destroy_listener_only_called_once_when_marked_as_destroyed)
{
    EXPECT_CALL(listener, callback()).Times(1);

    tracker.mark_destroyed();
}

TEST(LifetimeTrackerTest, is_still_alive_when_destroy_listener_called)
{
    mw::Weak<MockTracker> weak{&tracker};

    EXPECT_CALL(listener, callback()).Times(1)
        .WillOnce(InvokeWithoutArgs([&]{ EXPECT_THAT(weak, IsTrue()); }));
}
```

Reply:
Moved tracker and listener declarations to a test class, but I don't consider adding the listener to be part of the setup (some tests require multiple listeners to be added, some require holding on to the returned ID, etc. More obvious to just have each test add it itself however it needs.